File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.LinkedList; 
import java.util.Queue; 


class Solution {
    public int solution(int[] priorities, int location) {
        Queue<int[]> queue = new LinkedList<>(); 
        int answer = 0;
        
        for (int i = 0; i < priorities.length; i++) {
            queue.add(new int[]{priorities[i], i});
        }
        
        while(!queue.isEmpty()){
            int[] cur = queue.poll();
            int prior = cur[0];
            int index = cur[1];
            boolean hasMorePrior = false;
            
            for(int[] q : queue) {
                if(q[0] > prior) {
                    hasMorePrior = true;
                    break;
                }
            }
            
            if(hasMorePrior){
                queue.add(cur);
            } else {
                answer++;
                if(index == location) return answer;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/42587. 프로세스/프로세스.js
function solution(priorities, location) {
  const queue = priorities.map((p, i) => ({ priority: p, index: i }));
  const maxHeap = new PriorityQueue();

  priorities.forEach((p, i) => maxHeap.push(p, i));

  let answer = 0;
    
  while (queue.length) {
    const curr = queue.shift();
    if (curr.priority === maxHeap.peek()) {
      maxHeap.pop();
      answer++;
      if (curr.index === location) return answer;
    } else {
      queue.push(curr);
    }
  }
  return -1;
}

class PriorityQueue {
  constructor() {
    this.heap = [];
  }

  push(priority, index) {
    this.heap.push({ priority, index });
    this.heap.sort((a, b) => b.priority - a.priority);
  }

  pop() {
    return this.heap.shift();
  }

  peek() {
    return this.heap[0]?.priority || 0;
  }

  size() {
    return this.heap.length;
  }

  isEmpty() {
    return this.heap.length === 0;
  }
}

